packages/wallet/api/src/marshal-contexts.js

```diff
@@ -0,0 +1,362 @@
+// @ts-check
+import { makeScalarMap } from '@agoric/store';
+import { Far, makeMarshal, Remotable } from '@endo/marshal';
+import { HandledPromise } from '@endo/eventual-send'; // TODO: convince tsc this isn't needed
+
+const { details: X, quote: q } = assert;
+
+/**
+ * For a value with a known id in the board, we can use
+ * that board id as a slot to preserve identity when marshaling.
+ *
+ * @typedef {`board${Digits}`} BoardId
+ */
+
+/**
+ * @param {unknown} specimen
+ * @returns {specimen is BoardId}
+ */
+const isBoardId = specimen => {
+  return typeof specimen === 'string' && !!specimen.match(/board[^:]/);
+};
+
+/**
+ * When marshaling a purse, payment, etc. we partition the slots
+ * using prefixes.
+ *
+ * @template T
+ * @typedef {`${string & keyof T}:${Digits}`} WalletSlot<T>
```

Comment:
bravo on the string template type. non-blocking suggestion to help document how this works,
```suggestion
 * @template {Record<string, IdTable<*,*>} T
 * @typedef {`${keyof T}:${Digits}`} WalletSlot<T>
```

Reply:
I integrated most of that suggestion. Constraining `T` that way helps not only documentation but also got rid of one or two `@ts-expect-error`s.

But getting rid of `string &` doesn't seem to fly.